staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go

```diff
@@ -75,30 +90,39 @@ func Compile(s *schema.Structural) ([]cel.Program, []error) {
 	opts = append(opts, cel.Declarations(propDecls...))
 	env, err = env.Extend(opts...)
 	if err != nil {
-		allErrors = append(allErrors, err)
-		return nil, allErrors
+		compilationResults.Error = err
+		return compilationResults
 	}
-	programs := make([]cel.Program, len(celRules))
+	compResults := make([]CompilationResult, len(celRules))
 	for i, rule := range celRules {
+		var compilationResult CompilationResult
+		compilationResult.RuleIndex = i
+		compilationResult.Rule = rule
+		var errors []error
 		if rule.Rule == "" {
-			allErrors = append(allErrors, fmt.Errorf("valid is not specified"))
-			return nil, allErrors
-		}
-		ast, issues := env.Compile(rule.Rule)
-		if issues != nil {
-			allErrors = append(allErrors, fmt.Errorf("compilation failed for valid: %v with message: %v", rule, issues.Err()))
+			errors = append(errors, fmt.Errorf("rule is not specified"))
+			compilationResult.Errors = errors
 		} else {
-			prog, err := env.Program(ast)
-			if err != nil {
-				allErrors = append(allErrors, fmt.Errorf("program instantiation failed for valid: %v with message: %v", rule, err))
+			ast, issues := env.Compile(rule.Rule)
+			if issues != nil {
```

Comment:
we are better off returning a single error with `issues.Err()` It will be more readable.

Reply:
WDYT of this?